Proposal for re-architecting apply

pwittrock opened the pull request.

mengqiy commented:
/assign

cblecker commented:
@kubernetes/sig-cli-proposals 

Also cc: @kubernetes/sig-apps-proposals and @kubernetes/helm-maintainers .. I know there is some disparity between the way a `helm upgrade` works and the way a `kubectl apply` does when modifying existing manifests. I wonder if this is an opportunity to more closely align them within the kubernetes ecosystem :). If not, I apologize for the spam.

shiywang commented:
/sub

smarterclayton commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,91 @@
+  and optionally be able to show this to a user, whereas this conflict may not be detect with a PATCH
+  operation.
+
+#### New approach
```
Why not just move apply to the server via a formal API?

pwittrock replied:
@bgrant0607 asked a similar question.

My hope as that the steps in this proposal will make moving the logic to an API easier by reducing the complexity of the existing solution and making it a modular package that does not depend on cobra or other kubectl libraries to diff & update objects.

I'd be open to exposing the re-architected libraries as an API.

anguslees replied:
Because that's how we got strategic merge :P   In the general case, the server _can't_ know whether particular operations should be "replace" or "merge", and trying to capture all that flexibility in the API has created exactly the sort of complexity that motivated this doc.  The alternative is just to do it client-side, where naturally there is more context available to be able to implement the exact semantics intended by the client.

bgrant0607 replied:
Any examples in the API schema where it shouldn't be obvious? In the few ordered lists we have (e.g., command, args), those need to be replaced. All other cases should be merged. In the case of no conflict between the configured state and the live state, a merge and a replacement should be the same.

smarterclayton commented:
Yeah library components are good.  I just wanted to see it mentioned in the
doc so when the "apply on server" proposal gets thrown together I can point
back to that line.

On Mon, Sep 11, 2017 at 3:36 PM, Phillip Wittrock <notifications@github.com>
wrote:

> *@pwittrock* commented on this pull request.
> ------------------------------
>
> In contributors/design-proposals/sig-cli/apply_v2.md
> <https://github.com/kubernetes/community/pull/1028#discussion_r138169636>:
>
> > +- Patch syntax is not human readable, making it hard for humans to evaluate the end state
> +  that will result from the patch.  `--dry-run` is nearly useless for understanding the
> +  end state that will result from apply, and implementing a diff operation has been exceedingly
> +  copmlex.
> +- Generating a patch to make a change is more complicated than simply making the change to
> +  the object read from the server.  The apply command reads the live object from the server
> +  and diffs it against the local configuration.  Building a patch containing special merge
> +  directives is a more complicated approach than modifying the object read from the server
> +  and updating it with a PUT.
> +- Patches provide weaker optimistic locking guarantees than PUT.  When encountering an optimistic
> +  lock failure in apply, it is easy to re-read the object and try to update it again.  If the
> +  apply command tries to modify a field that was modified since apply read the object, it will fail
> +  and optionally be able to show this to a user, whereas this conflict may not be detect with a PATCH
> +  operation.
> +
> +#### New approach
>
> @bgrant0607 <https://github.com/bgrant0607> asked a similar question.
>
> My hope as that the steps in this proposal will make moving the logic to
> an API easier by reducing the complexity of the existing solution and
> making it a modular package that does not depend on cobra or other kubectl
> libraries to diff & update objects.
>
> I'd be open to exposing the re-architected libraries as an API.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/1028#discussion_r138169636>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pw1_1djB_9GgUcmUOF9q9MQFrHpZks5shYu6gaJpZM4PQjIn>
> .
>

bgrant0607 commented:
Server-side apply issue is #17333

bgrant0607 commented:
Also relevant are uses of both apply and patch in the declarative app mgmt doc: https://goo.gl/T66ZcD

Umbrella apply issue: https://github.com/kubernetes/kubernetes/issues/35234

Original apply proposal: https://github.com/kubernetes/kubernetes/issues/1702

bgrant0607 commented:
Original original issue: https://github.com/kubernetes/kubernetes/issues/1178

bgrant0607 commented:
Previous umbrella issue with design description: https://github.com/kubernetes/kubernetes/issues/15894

anguslees commented:
+100.  I've been slowly building up evidence to justify why I need to implement a new client-side "patch" as part of [kubecfg](http://github.com/ksonnet/kubecfg).  I'm happy to see that other people are sharing my pain :)

liggitt commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,91 @@
+#### New approach
+
+- Read the live object
+- Compare the live object to last-recorded and local files
```
do you still plan to record last-applied as an annotation?
discussion of the versioning issues (apply v1alpha1, time passes, want to apply v1beta2) would be good as well

pwittrock replied:
Yes, this proposal does not change what is stored for any of the objects, and works exactly the same with the exception that it uses PUT instead of PATCH to perform the update.  It might best to solve the issues around versioning in a server-side API.  I imagine the server side API would type convert all 3 objects (recorded from annotation, local from request and remote from live-lookup) to the same version before performing the diff-merge-update.

liggitt commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,91 @@
+
+- Read the live object
+- Compare the live object to last-recorded and local files
+- Update the fields on the live object that was read
```
using what merge strategies?

pwittrock replied:
The merge strategy / behavior would be exactly the same as the current version of apply, though more easily extended due to code structuring.

pwittrock commented:
@smarterclayton Good point.  I will add a section about that in the doc and the opportunity to integrate from the server-side on the future.

bgrant0607 commented:
Also relevant: discriminated unions: https://github.com/kubernetes/kubernetes/issues/35345

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,94 @@
+  version skew, as the client will depend on the server supporting a patch semantic.  There
+  is currently no versioning of the patch semantics, so the client cannot reliably know what
+  patch semantics the server supports.
+- Patch syntax is not human readable, making it hard for humans to evaluate the end state
```
Could you expand on the readability? Is this discussing strategic merge patch? Is there special syntax in the typical case other than for deletions? Where has the deletion syntax been made unreadable?

pwittrock replied:
Yes this is strategic merge patch.  I think there are 2 main problems, there are several different directives now (delete, delete-primitive, ordering), and how the patch will be applied is communicated out of band.  There are still a number of cases that are not supported by strategic patch that may require additional directives to support if done client side (e.g. how do we change the merge key for a field)

Directives:
- deletion of complex objects from lists
- deletion of primitive objects from lists (looks different)
- absolute ordering of items in a list

Additional issues with readability:
- merge keys for lists are not contained in the patch itself - unclear from looking at a patch whether it is adding an item to a list that looks similar to another item, or if it is updating the existing item in a list
- merge strategy not clear - are items in a list being replace, or are they being merged, (or using retain keys)?
- server silently drops directives that it doesn't recognize, so they may or may-not be used by the server - e.g. ordering may be randomized even with a directive specifying the ordering

mengqiy replied:
`retainKeys` is also a directive that you didn't mentioned above.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,94 @@
+  merge directives to the patch format on the server, and invoking them from the client.
+- Having to update both client and server to fix bugs introduces challenges to supporting
+  version skew, as the client will depend on the server supporting a patch semantic.  There
+  is currently no versioning of the patch semantics, so the client cannot reliably know what
```
We would version it by adding new content types, which I believe we also do with proto encodings.

But discoverability of supported versions is a good point.

bgrant0607 replied:
Re. discoverability: It's possible to specify supported content types for each verb X resource via OpenAPI

bgrant0607 replied:
Please add something about versioning the patch API in the future.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,94 @@
+  is currently no versioning of the patch semantics, so the client cannot reliably know what
+  patch semantics the server supports.
+- Patch syntax is not human readable, making it hard for humans to evaluate the end state
+  that will result from the patch.  `--dry-run` is nearly useless for understanding the
```
Why is dry run useless? Because it is also buggy and incomplete? Because it is implemented in the client rather than in the server?

pwittrock replied:
It doesn't show the user what they really want to know - what will be the new object in the cluster after apply is run.  It somewhat shows the changes being made, but without the context of what they are being made on.

- I would expect apply users care less about the patch format itself, and care more about the object in the cluster that will result from the patch.
- The patch output obfuscates validation errors caused by the values set by patch conflicting with other values set on the live object missing from the config (e.g. rolloutStrategy for Deployment)
- The patch itself is buggy and broken cases are hard to see just by looking at a patch (e.g. container ports merge key being incorrect)
- The server may apply the patch differently than the client, so running the patch locally against the file may result in something different than what the server does (e.g. the server doesn't support optional directives because it is skewed)

pwittrock replied:
FWIW, I don't understand what apply's current patch shows me when I do `--dry-run -o json`.  It looks like it is just showing me the local object without applying the patch.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,94 @@
+  to update an object and frequently has to be expanded with new side-directives describing
+  how the server should apply the patch.  Updates are done in the form of adding new
+  merge directives to the patch format on the server, and invoking them from the client.
+- Having to update both client and server to fix bugs introduces challenges to supporting
```
Links to exemplary bugs?

bgrant0607 replied:
For me: links to bugs are below.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,94 @@
+- Generating a patch to make a change is more complicated than simply making the change to
+  the object read from the server.  The apply command reads the live object from the server
+  and diffs it against the local configuration.  Building a patch containing special merge
+  directives is a more complicated approach than modifying the object read from the server
```
True.

However:
* My impression is that most of the complexity is not in the directives but in the API schema and defaulting behaviors, as I mentioned in the [server-side apply proposal](https://docs.google.com/document/d/1G4L4Kdyfg80DDuVOXHY0UbFuqDO29kECnOacVlseiFw/edit#).
* The schema needs to be accurately communicated to the client from the server (e.g., via extended OpenAPI) for client-side patch operations in order to support API aggregation and CRD, which re-introduces versioning and skew issues until the syntax stabilizes, but actually would improve skew behavior once it does.
* We need to support these patch behaviors independently of apply.

pwittrock replied:
> My impression is that most of the complexity is not in the directives but in the API schema and defaulting behaviors, as I mentioned in the server-side apply proposal.

Constructing a set of directives to order and delete items from a list results in more complicated code than just deleting the objects from the list.

> The schema needs to be accurately communicated to the client from the server (e.g., via extended OpenAPI) for client-side patch operations in order to support API aggregation and CRD, which re-introduces versioning and skew issues until the syntax stabilizes, but actually would improve skew behavior once it does.

Agreed.  This proposal should also work independently of whether it is done on the client or server.  The focus of this proposal is on rewriting the 3-way diff code to be maintainable.  The inputs will be the 3 versions of the object and the output will be the new object.  

> We need to support these patch behaviors independently of apply.

Yes, but why do we need to use them in apply?

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,94 @@
+  and diffs it against the local configuration.  Building a patch containing special merge
+  directives is a more complicated approach than modifying the object read from the server
+  and updating it with a PUT.
+- Patches provide weaker optimistic locking guarantees than PUT.  When encountering an optimistic
```
I'm not sure that's the case. I thought resourceVersion could be provided in patches.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,94 @@
+- Compare the live object to last-recorded and local files
+- Update the fields on the live object that was read
+- Send a PUT to update the modified object
+- If encountering optimistic lock failure, retry after re-reading the object
```
Specifically, GET, merge, and then try to PUT again.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,94 @@
+testing challenges.  We should instead modularize discrete responsibilities - collating the
+object values and updating the target object.
+
+#### Phase 1: Parse last-recorded, local, live objects and collate
```
last-applied-configuration?

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,94 @@
+
+Bad:
+
+- Larger request sizes to the server than when sending a patch
```
We had also discussed sending the full configured resource decorated with deletion directives, which would have been slightly larger.

bgrant0607 commented:
Draft of server-side proposal: https://docs.google.com/document/d/1G4L4Kdyfg80DDuVOXHY0UbFuqDO29kECnOacVlseiFw/edit#

mengqiy commented:
> Using compiled-in API types in kubectl

This should be solved by https://github.com/kubernetes/kubernetes/pull/51321. But it didn't make it in 1.8, I will reopen it after code freeze.

mengqiy commented:
Link the related proposal of sending mergekeys in the patch body as a directive: https://github.com/kubernetes/community/pull/476
The proposal above is trying to solve the problem of the mergekeys between client and serve are not synced in the case of version skew. This problem will be resolved if apply (diff+merge) is done completely at server-side (or client-side).

pwittrock commented:
> Using compiled-in API types in kubectl

@mengqiy already has a PR for this that has been lgtm'ed and approved.  Should go once the code freeze is lifted.

> Client-side apply

This doesn't directly address this, but I think we will want to rewrite / restructure the code in addtion to moving it to the server.

pwittrock commented:
> I don't have a strong opinion about PATCH vs. PUT, but it seems to me that this doesn't solve what I see as the 2 biggest problems:

The biggest immediate problem I face with apply is that the code to generate a PATCH is complex and difficult to change / review.  (A contributor spent 1-2 releases trying to implement `diff` and the code was so complex we had to drop 2 separate attempts).  Moving it to the server will not fix this unless we rewrite it in the process.

bgrant0607 commented:
@pwittrock Could you please reference some specific issues and/or PRs that illustrate the general problems that need to be solved?

php-coder commented:
This proposal should be places into `community/contributors/design-proposals/cli` directory (because `sig-cli` was renamed recently).

pwittrock commented:
@bgrant0607 Yes, good suggestion.  I have a few in mind that I think will do a good job.  I will add a section to the doc that performs an analysis of past PRs, how they were written, and how they would have been written under the new architecture.

pwittrock commented:
@bgrant0607 Updated with a better explanation of why the patch isn't human readable, and several PRs that show the complexity of making changes to the diff logic + accompanying analysis of where the complexity comes from.

pwittrock commented:
@php-coder thanks.  will move once all comments have been addressed - don't want to risk losing them.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+- Support deletion of primitives from lists
+- Lines (non-test): ~200
+- ~6 weeks
+[44597](https://github.com/kubernetes/kubernetes/pull/44597/files)
```
This was the wrong solution to this problem, IMO.

mengqiy replied:
I want to understand why.

bgrant0607 replied:
We want the correct behavior regardless whether POST, PUT, PATCH (any flavor), or apply is used. Therefore, the proper fix would be in the API. My proposal in https://github.com/kubernetes/kubernetes/issues/35345 was to add discriminators. It was incorrectly concluded that "We should set discriminators by default", and was dismissed as unsafe without further explanation.

Regardless, unions/OneOf are a property of the API schema. A patch directive should not have been needed. The patch version could be specified via Content-Type, which is how we currently distinguish among the flavors of patches:
https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#patch-operations

bgrant0607 replied:
And it's not clear to me that changing default API behavior is any more or less a breaking change than changing default kubectl behavior with no way to preserve the previous behavior.

bgrant0607 replied:
And what would break, actually? A patch to another key in a union should result in validation failure in the case that the previously set key were not cleared. Making such a request succeed would not be a breaking change -- all requests that previously worked would still work the same way.

pwittrock replied:
I would be happy to see Union get implemented.  It will need a champion for that to happen.  As things stand today, don't see my team tackling this in the next year.

bgrant0607 replied:
Even without discriminated unions, I don't see why directives were necessary.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+- Keep ordering of items when merging lists
+- Lines (non-test): ~650
+[46161](https://github.com/kubernetes/kubernetes/pull/46161/files#diff-101008d96c4444a5813f7cb6b54aaff6)
+- Support using multiple fields for a merge key
```
This seems like it will still be a problem.

mengqiy replied:
Maybe we can introduce multi-fields merge key as a new tag for apply. It may be the right chance to fix the wrong merge key issue, e.g. container.ports.
Server-side apply in the future can reuse this new merge key.

pwittrock replied:
It will still be a problem for PATCH, but provides a simpler solution for apply.

We won't need to come up with a strategy for communicating the merge keys within the PATCH (either explicitly or implicitly through versioning).  A relatively simple solution would be communicate the updated merge-keys to the client in openAPI, and then the client can decide if they want to use these keys or not.

Solving this for PATCH is a larger issue.

bgrant0607 replied:
Patches shouldn't need to communicate merge keys. That should be a property of the schema, not a property of the patch.

pwittrock replied:
If it is not explicitly part of the patch, then we need to introduce versioning semantics so both the client and server agree on the implicit merge keys since they may change.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+
+### Analysis - causes of complexity
+
+Apply is implemented by diffing the 3 sources (last-applied, local, remote) as 2 2-way diffs and then
```
Alternative implementation that has been proposed in the past:
1. Diff last-applied and local, and generate a patch by supplementing local with list and map deletion directives
2. Diff last-applied and remote/live to determine whether the live state was changed in a conflicting manner

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+
+Additionally, the logic to apply the patch must be updated both on the the server side
+and on the client, but cannot be done so transactionally.
+Because there there is no way for the client to know what directives a server supports (may be old),
```
Regardless what we do for apply, it sounds like we need a discovery mechanism for this.

bgrant0607 replied:
If we use content types for specifying the patch version, that can be represented in OpenAPI.

pwittrock replied:
Yes that sounds like a good direction.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+new directives must always be present but the patch must continue to function if they are ignored
+by the server.
+Because the server has no guarantee the client will send all directives it expects (may be old), it must
+not require any new directives to apply a patch "correctly".
```
It sounds like the behavior before the new directives were implemented was also incorrect?

pwittrock replied:
Yes, but the directives may only be needed for some cases.  For example, apply works for merging `Container.Ports` when 2 Ports don't share the same port number with different protocols.  This should continue to work with the new directives.

bgrant0607 replied:
No directives should be needed for that case.

pwittrock replied:
The client needs to communicate what it thinks the merge keys are - either implicitly through some versioning, or explicitly as some directive - for us to change the merge keys for things.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+    beforehand)
+  - Patch hides values on the remote object, which may have hidden interactions with the patch values
+    (e.g. Deployment.Spec.RolloutStrategy)
+- Generating a patch to make a change is more complicated than simply making the change to
```
It seems to me that the client vs. server discussion isn't the main point. 

The main point is whether the merge is accomplished via separate diff and patch passes, or with a custom merge implementation.

bgrant0607 replied:
And the new implementation should be mostly reusable for a server-side apply API.

pwittrock replied:
Yes exactly.  I can change the name of the doc if that would help.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+  - are spread across 1000+ lines of existing code
+- Changes that add new directives require updates in multiple locations - create patch + apply patch
+
+### PRs
```
I want overlays to work, so I think all of these problems still need to be solved for strategic merge patch.

pwittrock replied:
Agreed.  We should talk offline about this, since the requirements for overlays may be a subset of what is need for apply to work.  The semantics of apply are a bit different (though there is a lot of overlap) since apply is merging 3 modified versions of the same object to produce an updated object.  Whereas overlays do work in this context.

bgrant0607 commented:
I'm not opposed to implementing a custom merge operation for apply if that's simpler than diff and patch, but I disagree that it would address most of the difficulties discussed in the motivation.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+- Introduce logic for new a merge strategy by defining a new *Visitor* implementation
+- Introduce logic on structuring of a field by updating the parsing function for that field type
+
+If the apply diff logic was redesigned, most of the preceding PRs could be implemented by
```
Note that this would still change user-visible behavior. While technically bug fixes, we should consider how to communicate such changes to users, whether we need to retain prior behavior, and how we should transition to implementations with different behavior.

pwittrock replied:
Yes this makes sense.

In some cases the fixes purely fix something that was broken or unspecified (e.g. maintaining ordering for lists when merging is strictly more correct than randomizing the ordering).  In other cases I think it is pretty clearly an improvement, but the change should be communicated (e.g. deleting an item from a primitive lists now deletes it).  

The cases more related to your point would be fixing an unintended / broken merge-key or merge-strategy.

pwittrock commented:
@bgrant0607 

> I'm not opposed to implementing a custom merge operation for apply if that's simpler than diff and patch, but I disagree that it would address most of the difficulties discussed in the motivation.

The motivation then didn't do a sufficient job of arguing that the current code structure is now unmaintainable and brittle.  This is due largely to how the needs have evolved, and the original structure does not support extending the logic.

There seems to be consensus with the @apelisse and @mengqiy  that the [new architecture](https://github.com/kubernetes/kubernetes/pull/52349) will be simpler to maintain.

pwittrock commented:
@bgrant0607 @smarterclayton I've tried to address your comments.  LMK if there is anything else.

pwittrock commented:
/assign @apelisse

pwittrock commented:
@apelisse LMK if there is anything I should change since you are reviewing the code.

bgrant0607 commented:
@pwittrock Did you intend to push new updates? I don't see any commits newer than about a week old

pwittrock commented:
@bgrant0607 I not sure what changes to this proposal you are requesting before it is merged.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+`kubectl apply` reads a file or set of files, and updates the cluster state based off the file contents.
+It does a couple things:
+
+1. Create / Update / (Delete) the resource based on the resources
```
the live resources based on the file contents

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+It does a couple things:
+
+1. Create / Update / (Delete) the resource based on the resources
+2. Only set / delete fields defined in the file either currently or in during last update from the file
```
Update currently and previously configured fields, without clobbering fields set by other means, such as imperative kubectl commands, other deployment and management tools, admission controllers, initializers, horizontal and vertical autoscalers, operators, and other controllers.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+1. Create / Update / (Delete) the resource based on the resources
+2. Only set / delete fields defined in the file either currently or in during last update from the file
+  - Retain updates to fields made by controllers that were never managed in the file
+3. Merge fields based on the field name, merge lists based on a field contained within the list elements
```
I would move this below. It's a detail of field updates.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+3. Merge fields based on the field name, merge lists based on a field contained within the list elements
+
+The apply logic has been difficult to maintain from a contributor perspective, and difficult
+to understand from a user perspective.  Much of this pain has comes from the `apply` command out
```
Much of this pain has come from trying to paper over API schema complexity in an ad hoc incremental fashion using imperative patch directives.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+The apply logic has been difficult to maintain from a contributor perspective, and difficult
+to understand from a user perspective.  Much of this pain has comes from the `apply` command out
+growing its implementation.  In order to continue to support apply and provide better user visibility
+into what it is doing, the command must be re-architected to break client dependencies on the server-side
```
I thought we agreed that client-server was not the main issue.

bgrant0607 replied:
In order to make apply sufficiently maintainable and extensible to new API types, as well as to make its behavior more intuitive for users, the merge behavior, including how it is specified in the API schema, must be systematically redesigned and more thoroughly tested.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+
+Issues with updating the object by calculating and sending a patch to the server
+
+- Patch API is not fully featured enough to represent all of the operations required
```
has been frequently expanded

It doesn't have to be this way.

bgrant0607 commented on contributors/design-proposals/sig-cli/apply_v2.md:
```diff
@@ -0,0 +1,201 @@
+- Patch API is not fully featured enough to represent all of the operations required
+  to update an object and frequently has to be expanded with new side-directives describing
+  how the server should apply the patch.  Updates are done in the form of adding new
+  merge directives to the patch format on the server, and invoking them from the client.
```
Again, please rephrase. The schema MUST come from the server, in order to properly work with version skew, API aggregation, CRDs, etc.

pwittrock commented:
@bgrant0607 I've updated the doc to be focussed only on the code refactoring.  I added some bugs and called out the need to fix issues with the schema semantics as a separate item.

smarterclayton commented on contributors/design-proposals/cli/apply_refactor.md:
```diff
@@ -0,0 +1,184 @@
+- Compare the live object to last-applied and local files
+- Update the fields on the live object that was read
+- Send a PUT to update the modified object
+- If encountering optimistic lock failure, retry after re-reading the object
```
retry back to step 1?

smarterclayton commented on contributors/design-proposals/cli/apply_refactor.md:
```diff
@@ -0,0 +1,184 @@
+
+In order to make apply sufficiently maintainable and extensible to new API types, as well as to make its
+behavior more intuitive for users, the merge behavior, including how it is specified in the API schema,
+must be systematically redesigned and more thoroughly tested.
```
+1

pwittrock replied:
That was Brian :)

smarterclayton commented:
No further comments from me

bgrant0607 commented:
Thanks.
/lgtm
